vizro-ai/src/vizro_ai/dashboard/graph/code_generation.py

```diff
@@ -0,0 +1,184 @@
+
+
+logger = logging.getLogger(__name__)
+
+
+Messages = List[BaseMessage]
+"""List of messages."""
+
+
+class GraphState(BaseModel):
+    """Represents the state of dashboard graph.
+
+    Attributes
+        messages : With user question, error messages, reasoning
+        dfs : Dataframes
+        df_metadata : Cleaned dataframe names and their metadata
+        dashboard_plan : Plan for the dashboard
+        pages : Vizro pages
+        dashboard : Vizro dashboard
+
+    """
+
+    messages: List[BaseMessage]
+    dfs: List[pd.DataFrame]
+    df_metadata: DfMetadata
+    dashboard_plan: DashboardPlanner = None
+    pages: Annotated[List, operator.add]
+    dashboard: vm.Dashboard = None
+
+    class Config:
+        """Pydantic configuration."""
+
+        arbitrary_types_allowed = True
+
+    @validator("dfs")
+    def check_dataframes(cls, v):
+        """Check if the dataframes are valid."""
+        if not isinstance(v, list):
+            raise ValueError("dfs must be a list")
+        for df in v:
+            if not isinstance(df, pd.DataFrame):
+                raise ValueError("Each element in dfs must be a Pandas DataFrame")
+        return v
+
+
+def _store_df_info(state: GraphState, config: RunnableConfig) -> Dict[str, DfMetadata]:
+    """Store information about the dataframes."""
+    dfs = state.dfs
+    df_metadata = state.df_metadata
+    messages = state.messages
+    current_df_names = []
+    with tqdm(total=len(dfs), desc="Store df info") as pbar:
+        for df in dfs:
+            df_schema, df_sample = _get_df_info(df)
+
+            llm = config["configurable"].get("model", None)
+            data_sum_chain = df_sum_prompt | llm.with_structured_output(DfInfo)
+
+            df_name = data_sum_chain.invoke(
+                {
+                    "messages": messages,
+                    "df_schema": df_schema,
+                    "df_sample": df_sample,
+                    "current_df_names": current_df_names,
+                }
+            ).dataset_name
+
+            current_df_names.append(df_name)
+
+            pbar.write(f"df_name: {df_name}")
+            pbar.update(1)
+            df_metadata.metadata[df_name] = DataFrameMetadata(df_schema=df_schema, df=df, df_sample=df_sample)
+
+    return {"df_metadata": df_metadata}
+
+
+def _dashboard_plan(state: GraphState, config: RunnableConfig) -> Dict[str, DashboardPlanner]:
+    """Generate a dashboard plan."""
+    node_desc = "Generate dashboard plan"
+    pbar = tqdm(total=2, desc=node_desc)
+    query = state.messages[0].content
+    df_metadata = state.df_metadata
+
+    llm = config["configurable"].get("model", None)
+
+    _execute_step(pbar, node_desc + " --> in progress", None)
+    dashboard_plan = _get_dashboard_plan(query=query, model=llm, df_metadata=df_metadata)
+    _execute_step(pbar, node_desc + " --> done", None)
+    pbar.close()
+
+    return {"dashboard_plan": dashboard_plan}
+
+
+class BuildPageState(BaseModel):
```

Comment:
Don't really understand this - what does this make different from the overall state?

Reply:
This is the individual state for each page build. Then the `operator.add` in `pages: Annotated[List, operator.add]` from the parent state help consolidate all results together.